### ISSUE TITLE
Speedup distribution() - [closed]

### ISSUE THREAD

**Issue by jaraco** [python/importlib_metadata#231]
In GitLab by [@anntzer](https://gitlab.com/anntzer) on Dec 8, 2019, 19:25

_Merges perf -> master_

1st commit:

This speeds up `importlib_metadata.distribution()` ~4-6x both on
near-empty venvs and on "well-populated" ones.

The avoidance of both regexes and Path objects is necessary to achieve
the speedup.

2nd commit:

Save another 25-50% in runtime (i.e. up to ~2x) by constructing fewer
Path objects.

3rd commit:

Save a call to isfile() and an attempt to build a zipp.Path.

-----

See discussion at https://gitlab.com/python-devs/importlib_metadata/issues/95.

Context.pattern is now unused (it's too tied to regexes, which we don't use anymore).  Not sure if you want to just delete it or have it go through deprecation or something else; in the meantime I just added a nocover pragma.

It's a bit of a pity to have to go back to old-style string-based manipulations, but the speed difference is huge (around an order of magnitude).

**Comment by jaraco** [comment]
In GitLab by [@anntzer](https://gitlab.com/anntzer) on Dec 9, 2019, 24:39

added 1 commit

<ul><li>9f709bd9 - Less paths, more speed.</li></ul>

[Compare with previous version](/python-devs/importlib_metadata/merge_requests/103/diffs?diff_id=66617535&start_sha=184ccbb698b60aaa21a1194469deff76832318b0)

**Comment by jaraco** [comment]
In GitLab by [@anntzer](https://gitlab.com/anntzer) on Dec 9, 2019, 24:39

changed the description

**Comment by jaraco** [comment]
In GitLab by [@anntzer](https://gitlab.com/anntzer) on Dec 9, 2019, 07:25

added 1 commit

<ul><li>e54a040b - Save a call to isfile() and an attempt to build a zipp.Path.</li></ul>

[Compare with previous version](/python-devs/importlib_metadata/merge_requests/103/diffs?diff_id=66637551&start_sha=9f709bd9f251c29766ed8de5650dd09215edc103)

**Comment by jaraco** [comment]
In GitLab by [@codecov](https://gitlab.com/codecov) on Dec 9, 2019, 07:25

# [Codecov](https://codecov.io/gl/python-devs/importlib_metadata/pull/103?src=pr&el=h1) Report
> :exclamation: No coverage uploaded for pull request base (`master@d9a5f95`). [Click here to learn what that means](https://docs.codecov.io/docs/error-reference#section-missing-base-commit).
> The diff coverage is `100%`.

[![Impacted file tree graph](https://codecov.io/gl/python-devs/importlib_metadata/pull/103/graphs/tree.svg?width=650&token=UUI3tWw4Ex&height=150&src=pr)](https://codecov.io/gl/python-devs/importlib_metadata/pull/103?src=pr&el=tree)

```diff
@@           Coverage Diff           @@
##             master   #103   +/-   ##
=======================================
  Coverage          ?   100%           
=======================================
  Files             ?      2           
  Lines             ?    246           
  Branches          ?     21           
=======================================
  Hits              ?    246           
  Misses            ?      0           
  Partials          ?      0
```


| [Impacted Files](https://codecov.io/gl/python-devs/importlib_metadata/pull/103?src=pr&el=tree) | Coverage Δ | |
|---|---|---|
| [importlib\_metadata/\_compat.py](https://codecov.io/gl/python-devs/importlib_metadata/pull/103/diff?src=pr&el=tree#diff-aW1wb3J0bGliX21ldGFkYXRhL19jb21wYXQucHk=) | `100% <ø> (ø)` | |
| [importlib\_metadata/\_\_init\_\_.py](https://codecov.io/gl/python-devs/importlib_metadata/pull/103/diff?src=pr&el=tree#diff-aW1wb3J0bGliX21ldGFkYXRhL19faW5pdF9fLnB5) | `100% <100%> (ø)` | |

------

[Continue to review full report at Codecov](https://codecov.io/gl/python-devs/importlib_metadata/pull/103?src=pr&el=continue).
> **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta)
> `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> Powered by [Codecov](https://codecov.io/gl/python-devs/importlib_metadata/pull/103?src=pr&el=footer). Last update [d9a5f95...695c976](https://codecov.io/gl/python-devs/importlib_metadata/pull/103?src=pr&el=lastupdated). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments).

**Comment by jaraco** [comment]
In GitLab by [@jaraco](https://gitlab.com/jaraco) on Jan 8, 2020, 02:49

Thanks for this effort and thanks for your patience while I mull over this MR. Overall, I'm impressed that you've been able to squeeze out such an improvement with such a modest change.

I have a few reservations that I'll try to describe here.

## Untested

This test is untested, so would be fairly easy to be lost in a subsequent refactoring. It's not enough to have a comment at the beginning showing it's micro-optimized, as someone else could come along and make readability optimizations or other refactorings that would undo the benefits brought by this change. Can you conceive of a test that would adequately capture the expectation we're trying to meet here, something that would fail before the change and pass without it? I realize this ask is difficult for performance-bound changes like this one. Perhaps if there were some way to measure syscalls or bytecode instructions run, that could be an environment-agnostic measure of the complexity.

## Correctness

I'm not sure the "finding files at the root of a zipfile behavior" (lines 431-432) are correct. For example, is `os.path` the right separator (seems it is)? It does feel a little sloppy to create a list of children with lots of duplicates (for every grandchild).

Also, exceptions on `listdir(root)` would have bubbled up to the caller. Now they'll be silently ignored.

## Complexity

This change conflates a number of distinct behaviors into a single, long function. The semantic complexity of this function is much greater than that of any of the constituent functions it's replacing.

## Conflated Concerns

Relating to the prior concern, this change causes concerns to be interwoven. For example consider the "legacy" support for EGG-INFO files. In the prior implementation, that behavior is nicely encapsulated in the `_matches_legacy` behavior, and it's pretty obvious how that behavior could be removed if support for eggs was no longer needed. With this new implementation, one has to consider the interactions of those lines with all of the other (otherwise unrelated) concerns.

## Alternatives

Could there be an alternative, less invasive approach that retains more of the current structure, such as only optimizing `.iterdir()` and regex matching?

I considered that maybe memoizing one of these methods would have comparable performance benefits, but I don't think it would, as the performance bottleneck exists primarily with the first call.

----

My main concern is the lack of tested assurance with this change. Can you conceive of a way to address that?

**Comment by jaraco** [comment]
In GitLab by [@anntzer](https://gitlab.com/anntzer) on Jan 8, 2020, 10:53

added 1 commit

<ul><li>b24c04b1 - Address review comments.</li></ul>

[Compare with previous version](/python-devs/importlib_metadata/merge_requests/103/diffs?diff_id=69951049&start_sha=e54a040b81bf42ea0e79eb8375ddfa853d1cecc7)

**Comment by jaraco** [comment]
In GitLab by [@anntzer](https://gitlab.com/anntzer) on Jan 8, 2020, 10:53

# Untested

I think what makes the new version fast is clearly documented with a comment (no regexes, use strs and not Paths).  I can even extend the comment to something like "if you are going to change this, please make sure to check the perf on this-or-that".  I think an automated test for that is not realistic (even if measuring syscalls or bytecode instructions was doable, it would be a much larger project to implement).  Compare with e.g. https://github.com/python/cpython/pull/15134 which is likewise a perf-only PR and was merged solely on the basis of reported timings.

# Correctness

re: os.path: zipfile always uses forward slashes (https://github.com/python/cpython/blob/8d62df60d8733d0fa9aee14ef746d0009a7a9726/Lib/zipfile.py#L350) and has likely done so since a long time ago (https://bugs.python.org/issue440693) so the new version (splitting on `/` on posix, `\/` on win) is arguably more correct than the previous _matches_legacy (always splitting on `\/`).  I guess I can even replace `os.path.split` with just `.split('/')` or `posixpath.split` for even more correctness.  (Now done.)

re: creating lots of children: you have to iterate over the namelist anyways (whether explicitly like here, or implicitly using iterdir())...

re: other exceptions bubbling: I can probably replace the outer `except Exception` by `except NotADirectoryError` (well, except for Py2 support but that's doable), then I think it matches the earlier `is_dir()` check.  (Now done.)

# Complexity / Conflated Concerns

I can add some comments `# matches info` / `# matches legacy`.  (Now done.)

# Alternatives

Basic profiling (I use py-spy, but I'm sure other tools would work too) shows that the main costs of the old version are in *creating the Path objects* and *compiling the regexes*, so I doubt your suggetions will help.

Some kind of memoization (storing all dist/egg-infos found while looking for the one of interest, and resuming the iterator later if needed) may help with later calls, but not with the first one, and would add more complexity to this function.

To be clear, I agree the new version is more complex and less well architected, but this has the potential of being a *very* widely used function (for fetching `__version__` from packaging metadata instead of duplicating it), and the gain in performance is very large compared to the additional complexity.

**Comment by jaraco** [comment]
In GitLab by [@anntzer](https://gitlab.com/anntzer) on Jan 8, 2020, 11:10

added 1 commit

<ul><li>695c976c - Address review comments.</li></ul>

[Compare with previous version](/python-devs/importlib_metadata/merge_requests/103/diffs?diff_id=69953796&start_sha=b24c04b147099a99204fe05bf2ae8dfc1f78fda7)

**Comment by jaraco** [comment]
In GitLab by [@anntzer](https://gitlab.com/anntzer) on Jan 8, 2020, 13:58

added 1 commit

<ul><li>1b6c7dd1 - Address review comments.</li></ul>

[Compare with previous version](/python-devs/importlib_metadata/merge_requests/103/diffs?diff_id=69979127&start_sha=695c976c5618e318fe552d95ce839fbe33c5ce84)

**Comment by jaraco** [comment]
In GitLab by [@cwillenbrock](https://gitlab.com/cwillenbrock) on Jan 9, 2020, 21:48

Please consider something along the lines of the following change to address #94
```
diff --git a/importlib_metadata/__init__.py b/importlib_metadata/__init__.py
index 5405a9c..4ea9665 100644
--- a/importlib_metadata/__init__.py
+++ b/importlib_metadata/__init__.py
@@ -427,6 +427,9 @@ class MetadataPathFinder(NullFinder, DistributionFinder):
             children = os.listdir(root)
             path_type = pathlib.Path
         except OSError as exc:
+            # Permission denied
+            if exc.errno == errno.EACCES:
+                return
             # FileNotFoundError, NotADirectoryError.
             if exc.errno not in [
                     errno.ENOENT, errno.ENOTDIR]:  # pragma: nocover
```

**Comment by jaraco** [comment]
In GitLab by [@anntzer](https://gitlab.com/anntzer) on Jan 9, 2020, 21:54

I can do that if the devs agree, but right now I'm just trying to get the previously existing behavior in :)

**Comment by jaraco** [comment]
In GitLab by [@jaraco](https://gitlab.com/jaraco) on Jan 9, 2020, 22:44

Right, let's keep separate concerns separate, and have a separate MR for 94.

**Comment by jaraco** [comment]
In GitLab by [@jaraco](https://gitlab.com/jaraco) on Jan 10, 2020, 09:47

Antony,

We're close on this one.

In the [bugfix/95-perf branch](https://gitlab.com/python-devs/importlib_metadata/tree/bugfix/95-perf), I've pushed your commits (except the last one; I think I liked it better before), and then made changes to disentangle the concerns while retaining the key performance optimizations. Would you pull those commits and reset your MR to that state, at which point I'd like your comments and I'll make a final review before merging.

Thanks!

**Comment by jaraco** [comment]
In GitLab by [@anntzer](https://gitlab.com/anntzer) on Jan 10, 2020, 10:23

I had a look at your branch.  Unfortunately, it looks like the (very reasonable) abstractions it uses already cause a more than 2x slowdown.
FTR, the timing method I use is
```
$ rm -rf /tmp/testenv; python -mvenv /tmp/testenv && source /tmp/testenv/bin/activate && PIP_CONFIG_FILE=/dev/null pip install -qqq ~/src/extern/importlib_metadata ipython && ipython --quick                               
Python 3.8.0 (default, Nov  6 2019, 21:49:08) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.11.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import importlib_metadata                                                                                   

In [2]: %timeit importlib_metadata.distribution("ipython")                                                          
563 µs ± 1.39 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```
(compared to ~226µs on my last commit)

I haven't actually checked but I guess the generator stacks (both from yield (from) and from the generator comprehensions) are the costly ones.

---

As for the last commit here ("Address review comments"), I think at least the switch from os.path.split to .split("/") needs to stay (in theory someone could craft a zip file with an entry of the form "blah.egg\oops.egg-info" which should not be misunderstood as "blah.egg/oops.egg-info" on posix?).

**Comment by jaraco** [comment]
In GitLab by [@jaraco](https://gitlab.com/jaraco) on Jan 10, 2020, 18:00

Antony: Thanks for sharing the command. I found I could simplify it into a single command using pip-run:

```
~ $ pip-run -q ipython ~/p/importlib_metadata -- -m timeit -s 'import importlib_metadata' -- 'importlib_metadata.distribution("ipython")'                                                                    
500 loops, best of 5: 741 usec per loop
```

I'll use this technique to compare your head to mine and see if I can't get most or all of the gains back.

**Comment by jaraco** [comment]
In GitLab by [@cwillenbrock](https://gitlab.com/cwillenbrock) on Jan 10, 2020, 20:43

Should we expect that changes to this module will break Python2.7 support?
```
>>> import importlib_metadata
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/lib/python/importlib_metadata/__init__.py", line 402
    yield from os.listdir(self.root or '')
             ^
SyntaxError: invalid syntax
```

**Comment by jaraco** [comment]
In GitLab by [@jaraco](https://gitlab.com/jaraco) on Jan 10, 2020, 21:09

In cff8de2, I've removed the generators (for performance). The test now runs in 294 µs, down from 364 on its parent commit:

```
1000 loops, best of 5: 294 usec per loop
```

**Comment by jaraco** [comment]
In GitLab by [@jaraco](https://gitlab.com/jaraco) on Jan 10, 2020, 21:10

> Should we expect that changes to this module will break Python2.7 support?

The release will include Python 2.7 support (until this project drops support for it).

**Comment by jaraco** [comment]
In GitLab by [@jaraco](https://gitlab.com/jaraco) on Jan 10, 2020, 22:10

mentioned in merge request !106

**Comment by jaraco** [comment]
In GitLab by [@jaraco](https://gitlab.com/jaraco) on Jan 10, 2020, 23:44

Antony, I'm going to move forward with !106, which incorporates most of the changes herein. Thanks again for the detailed work on this.

Please confirm the performance is adequate and feel free to send subsequent pull requests for additional optimizations.

**Comment by jaraco** [comment]
In GitLab by [@jaraco](https://gitlab.com/jaraco) on Jan 10, 2020, 23:44

closed

**Comment by jaraco** [comment]
In GitLab by [@anntzer](https://gitlab.com/anntzer) on Jan 13, 2020, 14:05

Locally I observe that !106 is around 2.5x slower than this PR (~940ms vs 380ms).
I think there are some relatively low-lying fruits that can close most of the gap:
```patch
diff --git i/importlib_metadata/__init__.py w/importlib_metadata/__init__.py
index a84cc6f..6884461 100644
--- i/importlib_metadata/__init__.py
+++ w/importlib_metadata/__init__.py
@@ -408,6 +408,7 @@ class FastPath:
 
     def __init__(self, root):
         self.root = root
+        self._root_n_low = os.path.split(root)[1].lower()
 
     def joinpath(self, child):
         return pathlib.Path(self.root, child)
@@ -430,10 +431,9 @@ class FastPath:
             )
 
     def is_egg(self, search):
-        root_n_low = os.path.split(self.root)[1].lower()
-
+        root_n_low = self._root_n_low
         return (
-            root_n_low == search.normalized + '.egg'
+            root_n_low == search.versionless_egg_name
             or root_n_low.startswith(search.prefix)
             and root_n_low.endswith('.egg'))
 
@@ -452,7 +452,7 @@ class Prepared:
     """
     A prepared search for metadata on a possibly-named package.
     """
-    normalized = ''
+    versionless_egg_name = ''
     prefix = ''
     suffixes = '.dist-info', '.egg-info'
     exact_matches = [''][:0]
@@ -461,10 +461,11 @@ class Prepared:
         self.name = name
         if name is None:
             return
-        self.normalized = name.lower().replace('-', '_')
-        self.prefix = self.normalized + '-'
+        normalized = name.lower().replace('-', '_')
+        self.versionless_egg_name = normalized + '.egg'
+        self.prefix = normalized + '-'
         self.exact_matches = [
-            self.normalized + suffix for suffix in self.suffixes]
+            normalized + suffix for suffix in self.suffixes]
 
 
 @install
```

Avoiding having to repeatedly call os.path.split in is_egg already cuts down the runtime to 540ms(!)... including ~10ms by loading root_n_low into a local variable rather than doing three attribute accesses.
Avoiding recomputing `self.normalized + '.egg'` in is_egg shaves another 20ms.

Also, adding `__slots__` to FastPath and Prepared would shave another ~10ms as well.

I think most of the remaining difference come down to having everything in local variables vs. having to create intermediate objects and is unlikely to be gapped without having a large function that does everything; I won't insist on it.

I can make a new PR or let you pick up the patch, whatever works for you.

TL;DR: os.path.split is really slow :-)

**Comment by jaraco** [comment]
In GitLab by [@anntzer](https://gitlab.com/anntzer) on Jan 20, 2020, 18:37

mentioned in merge request !109
